feat(ingest): add stateful ingestion to the ldap source

bda618 opened the pull request:
Add stateful ingestion to the corpuser for ldap source. 

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

aditya-radhakrishnan reviewed (dismissed):
Can you please add tests for this change? Please update both [integration tests](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/integration/ldap/test_ldap.py) and [unit tests](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/unit/test_ldap_source.py)

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/state/ldap_state.py:
```diff
@@ -0,0 +1,50 @@
+)
+
+
+class BaseLdapCheckpointState(
```
Could you rename this to just `LdapCheckpointState` without the `Base`?

bda618 replied:
Sure, will do

bda618 replied:
completed

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/ldap.py:
```diff
@@ -117,12 +140,9 @@ class LDAPSourceConfig(ConfigModel):
+class LDAPSourceReport(StaleEntityRemovalSourceReport):
     def report_dropped(self, dn: str) -> None:
-        self.dropped_dns.append(dn)
+        self.report_stale_entity_soft_deleted(dn)
```
Revert this change. Stale entity soft deletion is automatically reported from [here](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/state/stale_entity_removal_handler.py#L225)

bda618 replied:
competed

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/ldap.py:
```diff
@@ -365,6 +404,10 @@ def build_corp_user_mce(
             user_snapshot.aspects.append(GroupMembershipClass(groups=groups))
 
+        self.stale_entity_removal_handler.add_entity_to_state(
+            type="corpuser", urn=f"urn:li:corpuser:{ldap_user}"
```
please construct the `urn` using [this function](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/emitter/mce_builder.py#L170)

bda618 replied:
will do

bda618 replied:
completed

bda618 commented:
> Can you please add tests for this change? 
I am going to work on this, but it looks like I will need to setup framework first and make sure all current tests are working. They are currently failing at our enterpise because we have a pulling all libraries from a different internal repo and tests need to be updated to make them working.

bda618 commented:
There is still 1 failure related to kafka which I am not sure how to fix here.

bda618 commented:
Team, is there anything pending for this PR? I have found it was included Ito a last release - 9.0.1

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/state/ldap_state.py:
```diff
@@ -0,0 +1,48 @@
+
+    @classmethod
+    def get_supported_types(cls) -> List[str]:
+        return ["corpuser"]
```
For completeness, we should also support `corpGroup`

hsheth2 replied:
If it's easier to implement, you can take a look at the Tableau source for an example of how to make this CheckpointState much simpler

bda618 replied:
This has been completed

hsheth2 commented on metadata-ingestion/tests/integration/ldap/test_ldap.py:
```diff
@@ -97,3 +97,66 @@ def test_ldap_memberof_ingest(docker_compose_runner, pytestconfig, tmp_path, moc
+                                "type": "datahub",
+                                "config": {
+                                    "datahub_api": {
+                                        "server": "http://localhost:8080",
```
There isn't a live datahub instance running when the integration tests run, so this test will fail always fail in CI.

Given that much of this code is tested through other mechanisms, it's probably ok to not have an integration test for stateful ingestion.

bda618 replied:
Completed

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/ldap.py:
```diff
@@ -389,21 +437,27 @@ def build_corp_group_mce(self, attrs: dict) -> Optional[MetadataChangeEvent]:
+                    type="corpGroup", urn=builder.make_group_urn(full_name)
+                )
+
+            return MetadataChangeEvent(proposedSnapshot=group_snapshot)
```
you can simplify this file even more! instead of manually calling `add_entity_to_state` and creating the status aspects, you can use the helpers that we have. 

I know our docs aren't amazing at explaining these helpers, so apologies for that.

Example:

https://github.com/datahub-project/datahub/blob/308b4eae87b20fa1a557e78ac0cc0329f63e6e4e/metadata-ingestion/src/datahub/ingestion/source/tableau.py#L1612-L1616

bda618 replied:
Yes, np, I will check and make that update. I have another question related to def close(self):
    def close(self) -> None:
```
        """Closes the Source."""
        self.ldap_client.unbind()
        self.prepare_for_commit(). <---- not needed anymore
        super().close()
```
I believe with a recent update you made at the parent class I don't need    self.prepare_for_commit() anymore (it was required before)

hsheth2 replied:
Yup that's right - the super().close() call internally calls prepare_for_commit, so you don't have to do it again

bda618 replied:
> you can simplify this file even more! instead of manually calling `add_entity_to_state` and creating the status aspects, you can use the helpers that we have.
> 
> I know our docs aren't amazing at explaining these helpers, so apologies for that.
> 
> Example:
> 
> https://github.com/datahub-project/datahub/blob/308b4eae87b20fa1a557e78ac0cc0329f63e6e4e/metadata-ingestion/src/datahub/ingestion/source/tableau.py#L1612-L1616

I still need to complete this last request for the change

bda618 replied:
I am not bale to find these utilities (auto_status_aspect) at our corporate Datahub version 9.0.0. It is probably was added recently, but because we don't have it, I cannot even test this now.

hsheth2 replied:
Got it - yep it was merged Oct 13 and released as part of 0.9.1 a few weeks later

Note that it is possible to use newer cli versions with older DataHub releases, so you should be able to update to 0.9.1

bda618 replied:
I think I understand how to auto_status_aspect() works. Essentially, I pass in "work units" and I got back an updated "workunits" where for all entities  StatusClass(removed=False).

 So, I can remove 

 StatusClass(removed=False),

and use the following to add  StatusClass:

yield from auto_status_aspect(self.stale_entity_removal_handler.gen_removed_entity_workunits())

This part I have tested and updated. However, I am not sure about "add_entity_to_state" functionality. Could you please elaborate on this slight more?

hsheth2 replied:
Sure thing. The eventual code will should look like this:

https://github.com/datahub-project/datahub/blob/44cfd21a653e23f869526a3c1a83e11d51b42a86/metadata-ingestion/src/datahub/ingestion/source/tableau.py#L1616-L1619

Your understanding of `auto_status_aspect` is almost right. The one difference - you'd need to rename the existing `get_workunits` to `get_workunits_internal`, and create a new get_workunits method with the contents above that calls the get_workunits_internal one. This way, auto_status_aspect applies on everything produced by the rest of the source.

`auto_stale_entity_removal` is a similar automatic function - it calls add_entity_to_state for each entity, and then calls gen_removed_entity_workunits once everything has been produced. That way, you can remove all calls to add_entity_to_state and gen_removed_entity_workunits in the rest of your code.

hsheth2 reviewed (commented):
added one quick suggestion on how we can make this even simpler, but lgtm other than that!

bda618 commented:
@hsheth2, I need your help or advise on the potential issue I am trying to investigate. In short, we have found (by a hard way, at the prod) that some ldap accounts suddenly disappeared after the job run. It is important to note that we have a lot of accounts, few hundred thousands (!). I committed a set of tests and was able to reproduce the issue at my local environment (by using a subset of data). The issue is happening when I process a significant amount of accounts. I cannot confirm this yet, but I have feeling that it somehow related to number of URN (=accounts) that current checkpoint could keep in memory. This 5000 is what I see in  debug data window (however, this could be related to my debug tool limitation as well, so it is just a guess). Thus, state of the previous check point hold some account, say "aaa1", but it cannot be found at the current check point. However, the previous checkpoint was generated just 20 minutes ago, so account "aaa1" must be at the current checkpoint as well. We can make a reasonable guess that "aaa1" is just not lucky and was laid off in between, but the thing is that he is not the only one, there are hundred more. So, definitely something is not right. I need your help here. I know the code very well at this point, but such kind of issues needs a real expert.

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/ldap.py:
```diff
@@ -86,14 +98,27 @@ def set_cookie(
 
 
-class LDAPSourceConfig(ConfigModel):
+class LDAPSourceStatefulIngestionConfig(StatefulStaleMetadataRemovalConfig):
```
this entire class is no longer necessary!

bda618 replied:
I think I got it, I can remove  LDAPSourceStatefulIngestionConfig now and use StatefulStaleMetadataRemovalConfig instead.

hsheth2 reviewed (commented):
The code LGTM

In reviewing this PR, I actually found a bug in `auto_status_aspect` - the PR to fix it is here https://github.com/datahub-project/datahub/pull/6705.

I'm thinking that it probably makes sense to get that one merged in before merging this PR. That means it might take an extra day, and you need to update the .json files again.

bda618 commented:
> The code LGTM
> 
> In reviewing this PR, I actually found a bug in `auto_status_aspect` - the PR to fix it is here #6705.
> 
> I'm thinking that it probably makes sense to get that one merged in before merging this PR. That means it might take an extra day, and you need to update the .json files again.

I actually have found (before noticed your reply) that Status doesn't work, and fond this check at the helper class:
`isinstance(aspect, StatusClass)` 
In my case there is no such aspect at MetadataChangeEventClass, so logic was broken. I added Status manually and rolled back JSON files (they actually worked and found this issue). I also removed code for manual add to the state which I left behind by mistake. I tested all locally and it appears to be fine. So, essentially I need to know what to do with with those 2 Status lines (for corpuses and groups).

hsheth2 commented:
@bda618 I'm glad the changes from the other PR fixed the status stuff

LGTM = looks good to me :)

bda618 commented:
> @bda618 I'm glad the changes from the other PR fixed the status stuff
> 
> LGTM = looks good to me :)

Oh, that make sense now. 

I have failed tests because of that missing Status, but it is expected till PR is merged.

hsheth2 reviewed (approved):
LGTM

I've merged the fix to auto_status_aspect, but it looks like the tests are still failing

bda618 commented:
> LGTM
> 
> I've merged the fix to auto_status_aspect, but it looks like the tests are still failing

Your fix works fine in the real environment and fixed "status" aspect setup during an ingestion, I tested that. 
However, my integration test generates "AssertionError: MCEs differ" because of the missing:
`"com.linkedin.pegasus2avro.common.Status": {
                        "removed": false
 }`
So, I need to fix the test.

hsheth2 commented:
@bda618 usually `pytest tests/<whatever> --update-golden-files` will do the trick

bda618 commented:
> @bda618 usually `pytest tests/<whatever> --update-golden-files` will do the trick

Completed.

bda618 commented:
> @bda618 usually `pytest tests/<whatever> --update-golden-files` will do the trick

Completed

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/ldap.py:
```diff
@@ -370,6 +406,7 @@ def build_corp_user_mce(
     def build_corp_group_mce(self, attrs: dict) -> Optional[MetadataChangeEvent]:
         """Creates a MetadataChangeEvent for LDAP groups."""
         cn = attrs.get(self.config.group_attrs_map["urn"])
+        print(cn)
```
revert this change

bda618 replied:
I have made two  changes:
1) replaced  "self.report_stale_entity_soft_deleted(dn)" by  "self.dropped_dns.append(dn)"
2) removed "print(cn)"

I think my confusion was with the exact meaning of word "revert" in this context. Because there were so many changes at different times, so it was not clear to me what does "revert" mean? Should I revert to the original version of open source? Should I revert to my original version? Should I replace the code highlighted in "red" color by the code highlighted in "green" color as I usually do or should do exactly the opposite now because there is a word "revert" there?

I think just a simple replace this "..." by "..." because of "..." would be a way more easy to understand. Sorry, just some suggestions.

hsheth2 reviewed (changes requested):
Noticed two tiny things that I'd like to get resolved before merging

Sorry about the extra delay
